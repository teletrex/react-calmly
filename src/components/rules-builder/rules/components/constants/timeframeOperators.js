/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import { RULE_INPUT_TYPES } from '../../../constants';

const { NONE, RANGE, SINGLE_VALUE, WHEN_RELATION, WITHIN_PERIOD } = RULE_INPUT_TYPES;

export const TIMEFRAME_OPERATORS = [
  // Within:
  {
    inputType: WITHIN_PERIOD,
    label: 'within the last',
    value: 'WITHIN_THE_LAST',
  },

  // Today:
  {
    inputType: NONE,
    label: 'on today',
    value: 'ON_TODAY',
  },
  {
    inputType: NONE,
    label: 'before today',
    value: 'BEFORE_TODAY',
  },

  // Any time:
  {
    inputType: NONE,
    label: 'at any time',
    value: 'AT_ANY_TIME',
  },

  // Date:
  {
    inputType: SINGLE_VALUE,
    label: 'on',
    value: 'ON',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'after',
    value: 'AFTER',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'before',
    value: 'BEFORE',
  },

  // Date range:
  {
    inputType: RANGE,
    label: 'between',
    value: 'BETWEEN',
  },

  // When relation:
  {
    inputType: WHEN_RELATION,
    label: 'exactly',
    value: 'EXACTLY',
  },
  {
    inputType: WHEN_RELATION,
    label: 'more than',
    value: 'MORE_THAN',
  },
  {
    inputType: WHEN_RELATION,
    label: 'less than',
    value: 'LESS_THAN',
  },
];
