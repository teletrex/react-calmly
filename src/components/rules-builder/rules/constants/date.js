/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const DATE_PATTERN = 'd{1,2}/d{1,2}/d{4}';
