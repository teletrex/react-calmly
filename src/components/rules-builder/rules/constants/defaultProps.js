/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const DEFAULT_PROPS = {
  INPUT_FIELD_VALIDATION: [],
  OPERATORS: [],
  TIMEFRAME: {},
};
