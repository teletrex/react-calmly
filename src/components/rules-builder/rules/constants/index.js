/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export * from './date';
export * from './defaultProps';
export * from './periods';
export * from './propTypes';
export * from './ruleOperators';
export * from './validationMessages';
export * from './values';
export * from './whenRelations';
