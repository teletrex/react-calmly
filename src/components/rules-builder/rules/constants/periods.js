/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const PERIODS = [
  {
    label: 'days',
    value: 'DAYS',
  },
  {
    label: 'weeks',
    value: 'WEEKS',
  },
  {
    label: 'months',
    value: 'MONTHS',
  },
];
