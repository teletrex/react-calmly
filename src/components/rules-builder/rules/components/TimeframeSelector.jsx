/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React, { useState } from 'react';
import PropTypes from 'prop-types';

import { useTranslation } from '../../../../translation';
import { getInitialOperator, getPrefixedClasses } from '../helpers';
import { DEFAULT_PROPS, PROPTYPES, VALUES } from '../constants';
import { RULE_INPUT_TYPES } from '../../constants';

import TimeframeOperatorDropdown from './OperatorDropdown';
import { TIMEFRAME_OPERATORS } from './constants';
import { RangeDatePicker, SingleDatePicker, WhenRelation, Within } from './inputFields';

const getTimeframeOutput = (operator, value) => ({ operator, value });

const TimeframeSelector = ({ id: idPrefix, timeframe }) => {
  const { t } = useTranslation();
  const { onChange, data: { operator, value: initialValue } = {} } = timeframe;
  const id = `${idPrefix}--timeframe`;

  const initialTimeframeOperator = getInitialOperator(operator, TIMEFRAME_OPERATORS);

  const [timeframeValue, setTimeframeValue] = useState(initialValue);
  const [timeframeOperator, setTimeframeOperator] = useState(initialTimeframeOperator);

  const handleTimeframeOperatorChange = ({ selectedItem: newOperator }) => {
    const currentValue =
      newOperator.inputType === RULE_INPUT_TYPES.NONE ? VALUES.BLANK : VALUES.UNPROVIDED;

    setTimeframeValue(currentValue);
    setTimeframeOperator(newOperator);
    onChange(getTimeframeOutput(newOperator.value, currentValue));
  };

  const handleTimeframeValueChange = newValue => {
    setTimeframeValue(newValue);
    onChange(getTimeframeOutput(timeframeOperator.value, newValue));
  };

  const renderInputFields = () => {
    const commonProps = {
      onValueChange: handleTimeframeValueChange,
      id,
      operator: timeframeOperator,
      value: timeframeValue,
    };

    switch (timeframeOperator?.inputType) {
      case RULE_INPUT_TYPES.RANGE:
        return <RangeDatePicker {...commonProps} />;

      case RULE_INPUT_TYPES.SINGLE_VALUE:
        return <SingleDatePicker {...commonProps} />;

      case RULE_INPUT_TYPES.WHEN_RELATION:
        return <WhenRelation {...commonProps} />;

      case RULE_INPUT_TYPES.WITHIN_PERIOD:
        return <Within {...commonProps} />;

      default:
        return null;
    }
  };

  return (
    <div
      className={getPrefixedClasses('--timeframe')}
      data-testid="timeframe-selector"
      id={`${id}-container`}
    >
      <TimeframeOperatorDropdown
        id={id}
        label={t('select timeframe')}
        onOperatorChange={handleTimeframeOperatorChange}
        operator={timeframeOperator}
        operators={TIMEFRAME_OPERATORS}
      />
      {renderInputFields()}
    </div>
  );
};

TimeframeSelector.propTypes = {
  id: PropTypes.string.isRequired,
  timeframe: PROPTYPES.TIMEFRAME,
};

TimeframeSelector.defaultProps = {
  timeframe: DEFAULT_PROPS.TIMEFRAME,
};

export default TimeframeSelector;
