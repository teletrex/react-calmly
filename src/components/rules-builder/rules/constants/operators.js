/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import { RULE_INPUT_TYPES } from '../../constants';

const {
  MULTIPLE_VALUES,
  NONE,
  PERIOD_NUMBER,
  PERIOD_RANGE,
  RANGE,
  SINGLE_VALUE,
  WHEN_RELATION,
  WITHIN_PERIOD,
} = RULE_INPUT_TYPES;

// Common operators
export const BLANK_OPERATORS = [
  {
    inputType: NONE,
    label: 'is blank',
    value: 'IS_BLANK',
  },
  {
    inputType: NONE,
    label: 'is not blank',
    value: 'IS_NOT_BLANK',
  },
];

export const EXACT_OPERATORS = [
  {
    inputType: WHEN_RELATION,
    label: 'is exactly',
    value: 'IS_EXACTLY',
  },
  {
    inputType: WHEN_RELATION,
    label: 'is not exactly',
    value: 'IS_NOT_EXACTLY',
  },
];

export const IS_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is',
    value: 'IS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is not',
    value: 'IS_NOT',
  },
];

export const HAS_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'has',
    value: 'HAS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'has not',
    value: 'HAS_NOT',
  },
];

export const IS_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'is one of the following',
    value: 'IS_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'is not one of the following',
    value: 'IS_NOT_ONE_OF_THE_FOLLOWING',
  },
];

export const CONTAIN_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'contains',
    value: 'CONTAINS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'does not contain',
    value: 'DOES_NOT_CONTAIN',
  },
];

export const CONTAIN_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'contains one of the following',
    value: 'CONTAINS_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'does not contain one of the following',
    value: 'DOES_NOT_CONTAIN_ONE_OF_THE_FOLLOWING',
  },
];

export const RANGE_OPERATORS = [
  {
    inputType: RANGE,
    label: 'is between',
    value: 'IS_BETWEEN',
  },
];

// String related operators
export const STARTS_WITH_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'starts with',
    value: 'STARTS_WITH',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'does not start with',
    value: 'DOES_NOT_START_WITH',
  },
];

export const STARTS_WITH_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'starts with one of the following',
    value: 'STARTS_WITH_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'does not start with one of the following',
    value: 'DOES_NOT_START_WITH_ONE_OF_THE_FOLLOWING',
  },
];

export const ENDS_WITH_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'ends with',
    value: 'ENDS_WITH',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'does not end with',
    value: 'DOES_NOT_END_WITH',
  },
];

export const ENDS_WITH_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'ends with one of the following',
    value: 'ENDS_WITH_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'does not end with one of the following',
    value: 'DOES_NOT_END_WITH_ONE_OF_THE_FOLLOWING',
  },
];

// Number related operators
export const LESS_THAN_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is less than',
    value: 'IS_LESS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is less than or equal to',
    value: 'IS_LESS_OR_EQUAL',
  },
];

export const GREATER_THAN_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is greater than',
    value: 'IS_GREATER',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is greater than or equal to',
    value: 'IS_GREATER_OR_EQUAL',
  },
];

// Date related operators
export const TODAY_OPERATORS = [
  {
    inputType: NONE,
    label: 'is today',
    value: 'IS_TODAY',
  },
  {
    inputType: NONE,
    label: 'is after today',
    value: 'IS_AFTER_TODAY',
  },
  {
    inputType: NONE,
    label: 'is not after today',
    value: 'IS_NOT_AFTER_TODAY',
  },
  {
    inputType: NONE,
    label: 'is before today',
    value: 'IS_BEFORE_TODAY',
  },
  {
    inputType: NONE,
    label: 'is not before today',
    value: 'IS_NOT_BEFORE_TODAY',
  },
  {
    inputType: NONE,
    label: 'anniversary is today',
    value: 'ANNIVERSARY_IS_TODAY',
  },
];

export const ANNIVERSARY_OPERATORS = [
  {
    inputType: WHEN_RELATION,
    label: 'anniversary is',
    value: 'ANNIVERSARY_IS',
  },
];

export const WITHIN_OPERATORS = [
  {
    inputType: WITHIN_PERIOD,
    label: 'is within the last',
    value: 'IS_WITHIN_THE_LAST',
  },
  {
    inputType: WITHIN_PERIOD,
    label: 'is not within the last',
    value: 'IS_NOT_WITHIN_THE_LAST',
  },
  {
    inputType: WITHIN_PERIOD,
    label: 'is within the next',
    value: 'IS_WITHIN_THE_NEXT',
  },
  {
    inputType: WITHIN_PERIOD,
    label: 'is not within the next',
    value: 'IS_NOT_WITHIN_THE_NEXT',
  },
];

export const WITHIN_RANGE_OPERATORS = [
  {
    inputType: PERIOD_RANGE,
    label: 'is within the range of the last',
    value: 'IS_WITHIN_THE_RANGE_OF_THE_LAST',
  },
];

export const PERIOD_EQUAL_OPERATORS = [
  {
    inputType: PERIOD_NUMBER,
    label: 'year is',
    value: 'YEAR_IS',
  },
  {
    inputType: PERIOD_NUMBER,
    label: 'month is',
    value: 'MONTH_IS',
  },
  {
    inputType: PERIOD_NUMBER,
    label: 'day is',
    value: 'DAY_IS',
  },
];

export const BEFORE_AFTER_DATE_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is after',
    value: 'IS_AFTER',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is before',
    value: 'IS_BEFORE',
  }
]

export const DATA_AGGREGATION_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'by',
    value: 'BY',
  },

];
