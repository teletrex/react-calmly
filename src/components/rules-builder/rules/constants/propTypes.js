/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import PropTypes from 'prop-types';

import { RULE_INPUT_TYPES } from '../../constants';

const { any, arrayOf, func, oneOf, oneOfType, shape, string } = PropTypes;

const OPERATOR = oneOfType([
  string,
  shape({
    inputType: oneOf(Object.values(RULE_INPUT_TYPES)).isRequired,
    label: string.isRequired,
    value: string.isRequired,
  }),
]);
const OPERATORS = arrayOf(OPERATOR);

const TIMEFRAME = shape({
  data: shape({
    operator: string,
    value: any,
  }),
  onChange: func,
});

const OPERATOR_NAME = string;

const SELECT_OPTION = shape({
  label: string.isRequired,
  value: string.isRequired,
});

const SELECT_OPTIONS = arrayOf(SELECT_OPTION);

export const PROPTYPES = {
  OPERATOR,
  OPERATORS,
  OPERATOR_NAME,
  SELECT_OPTION,
  SELECT_OPTIONS,
  TIMEFRAME,
};
