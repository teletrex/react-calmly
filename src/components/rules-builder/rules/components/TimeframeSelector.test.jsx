/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
import { fireEvent, render, waitFor, within } from '@testing-library/react';
import find from 'lodash/find';
import userEvent from '@testing-library/user-event';

import { VALUES } from '../constants';
import { renderWithContext } from '../../utils/testHelpers';

import TimeframeSelector from './TimeframeSelector';
import { TIMEFRAME_OPERATORS } from './constants';

const timeframeData = {
  operator: 'EXACTLY',
  value: {
    value: 7,
    period: 'DAYS',
    whenRelation: 'AWAY',
  },
};
const timeframeOnChange = jest.fn();

const operatorTestCases = [
  ['input-field-range-date-picker', 'BETWEEN', [1640991600000, 1643410800000]],
  ['input-field-single-date-picker', 'AFTER', 1640991600000],
  ['input-field-when-relation', 'MORE_THAN', timeframeData.value],
  ['input-field-within', 'WITHIN_THE_LAST', { period: 'WEEKS', value: 3 }],
];

const defaultProps = {
  id: 'TestTimeframeSelectorElement',
  timeframe: {
    onChange: timeframeOnChange,
    data: timeframeData,
  },
};

describe('TimeframeSelector', () => {
  beforeEach(() => {
    timeframeOnChange.mockReset();
  });

  it('should render correctly', () => {
    const { container } = render(<TimeframeSelector {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly for read-only', () => {
    const { container } = renderWithContext(<TimeframeSelector {...defaultProps} />, {
      isReadOnly: true,
    });
    expect(container).toMatchSnapshot();
  });

  it('should set the timeframe value when a value is changed and input field is blurred', async () => {
    const { getByPlaceholderText } = render(<TimeframeSelector {...defaultProps} />);
    const dayInput = getByPlaceholderText('Enter value');

    fireEvent.change(dayInput, { target: { value: 9 } });
    await waitFor(() => expect(defaultProps.timeframe.onChange).toHaveBeenCalledTimes(0));

    fireEvent.blur(dayInput);

    const expectedValue = {
      ...defaultProps.timeframe.data,
      value: {
        ...defaultProps.timeframe.data.value,
        value: 9,
      },
    };

    await waitFor(() =>
      expect(defaultProps.timeframe.onChange).toHaveBeenCalledWith(expectedValue)
    );
  });

  it('should set the timeframe value when a timeframe operator is changed', () => {
    const { getAllByRole, getByText } = render(<TimeframeSelector {...defaultProps} />);
    const newOperator = find(TIMEFRAME_OPERATORS, { value: 'AT_ANY_TIME' });

    // expand the operator dropdown:
    const [operatorDropdown] = getAllByRole('listbox');
    const operatorDropdownButton = within(operatorDropdown).getByRole('button');
    userEvent.click(operatorDropdownButton);

    // select a new operator:
    const newOperatorItem = getByText(newOperator.label);
    userEvent.click(newOperatorItem);

    const expectedValue = {
      operator: newOperator.value,
      value: VALUES.BLANK,
    };

    expect(defaultProps.timeframe.onChange).toHaveBeenCalledWith(expectedValue);
  });

  test.each(operatorTestCases)(
    'should render a proper component when %p operator is selected',
    (testId, operatorValue, value) => {
      const props = {
        ...defaultProps,
        timeframe: {
          ...defaultProps.timeframe,
          data: {
            value,
            operator: operatorValue,
          },
        },
      };

      const { getByTestId } = render(<TimeframeSelector {...props} />);
      expect(getByTestId(testId)).toBeInTheDocument();
    }
  );
});
