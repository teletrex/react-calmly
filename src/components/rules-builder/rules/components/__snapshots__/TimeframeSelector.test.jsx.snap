// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeframeSelector should render correctly 1`] = `
<div>
  <div
    class="cds--timeframe"
    data-testid="timeframe-selector"
    id="TestTimeframeSelectorElement--timeframe-container"
  >
    <div
      class="cds--proposal-dropdown-wrapper cds--proposal-dropdown-wrapper--inline"
    >
      <div
        class="cds--proposal-dropdown"
        role="listbox"
      >
        <span>
          <div
            aria-label="select timeframe"
            class="cds--proposal-dropdown__field"
            role="button"
            tabindex="0"
          >
            <span
              class="cds--proposal-dropdown__field__label"
            >
              exactly
            </span>
            <div
              class="cds--proposal-dropdown__field__menu-icon"
            >
              <svg
                aria-hidden="true"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                style="will-change: transform;"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 11L3 6l.7-.7L8 9.6l4.3-4.3.7.7z"
                />
              </svg>
            </div>
          </div>
        </span>
      </div>
    </div>
    <div
      class="cds--input-field-container"
      data-testid="input-field-when-relation"
    >
      <div
        class="cds--number-input-wrapper"
      >
        <div
          class="cds--form-item"
        >
          <div
            class="cds--number cds--number--helpertext cds--number-input-wrapper_carbon-input"
          >
            <div
              class="cds--number__input-wrapper"
            >
              <input
                aria-label="Numeric input field with increment and decrement buttons"
                id="TestTimeframeSelectorElement--timeframe"
                max="Infinity"
                min="0"
                name="TestTimeframeSelectorElement--timeframe"
                pattern="[0-9]*"
                placeholder="Enter value"
                step="1"
                type="number"
                value="7"
              />
              <div
                class="cds--number__controls"
              >
                <button
                  aria-atomic="true"
                  aria-label="Increment number. Press Space or Enter in order to use icon."
                  aria-live="polite"
                  class="cds--number__control-btn up-icon"
                  focusable="false"
                  tabindex="-1"
                  title="Increment number. Press Space or Enter in order to use icon."
                  type="button"
                  unselectable="on"
                >
                  <svg
                    aria-hidden="true"
                    class="up-icon"
                    focusable="false"
                    height="4"
                    preserveAspectRatio="xMidYMid meet"
                    style="will-change: transform;"
                    viewBox="0 0 8 4"
                    width="8"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 4l4-4 4 4z"
                    />
                  </svg>
                </button>
                <button
                  aria-atomic="true"
                  aria-label="Decrement number. Press Space or Enter in order to use icon."
                  aria-live="polite"
                  class="cds--number__control-btn down-icon"
                  focusable="false"
                  tabindex="-1"
                  title="Decrement number. Press Space or Enter in order to use icon."
                  type="button"
                  unselectable="on"
                >
                  <svg
                    aria-hidden="true"
                    class="down-icon"
                    focusable="false"
                    height="4"
                    preserveAspectRatio="xMidYMid meet"
                    style="will-change: transform;"
                    viewBox="0 0 8 4"
                    width="8"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M8 0L4 4 0 0z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>

      </div>
      <div
        class="cds--proposal-dropdown-wrapper"
      >
        <div
          class="cds--proposal-dropdown"
          role="listbox"
        >
          <span>
            <div
              aria-label="Select period"
              class="cds--proposal-dropdown__field"
              role="button"
              tabindex="0"
            >
              <span
                class="cds--proposal-dropdown__field__label"
              >
                days
              </span>
              <div
                class="cds--proposal-dropdown__field__menu-icon"
              >
                <svg
                  aria-hidden="true"
                  focusable="false"
                  height="16"
                  preserveAspectRatio="xMidYMid meet"
                  style="will-change: transform;"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 11L3 6l.7-.7L8 9.6l4.3-4.3.7.7z"
                  />
                </svg>
              </div>
            </div>
          </span>
        </div>
      </div>
      <div
        class="cds--proposal-dropdown-wrapper"
      >
        <div
          class="cds--proposal-dropdown"
          role="listbox"
        >
          <span>
            <div
              aria-label="Select..."
              class="cds--proposal-dropdown__field"
              role="button"
              tabindex="0"
            >
              <span
                class="cds--proposal-dropdown__field__label"
              >
                away
              </span>
              <div
                class="cds--proposal-dropdown__field__menu-icon"
              >
                <svg
                  aria-hidden="true"
                  focusable="false"
                  height="16"
                  preserveAspectRatio="xMidYMid meet"
                  style="will-change: transform;"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 11L3 6l.7-.7L8 9.6l4.3-4.3.7.7z"
                  />
                </svg>
              </div>
            </div>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TimeframeSelector should render correctly for read-only 1`] = `
<div>
  <div
    class="cds--timeframe"
    data-testid="timeframe-selector"
    id="TestTimeframeSelectorElement--timeframe-container"
  >
    <span
      class="cds--text-space-left"
    >
      exactly
    </span>
    <span
      class="cds--rule-value-readonly"
    >
      <span
        class="cds--text-space-left"
      >
        7
      </span>
      <span
        class="cds--text-space-left"
      >
        days
      </span>
      <span
        class="cds--text-space-left"
      >
        away
      </span>
    </span>
  </div>
</div>
`;
